ssas/service/blacklist/blacklist.go

```diff
@@ -0,0 +1,325 @@
+		return fmt.Errorf(fmt.Sprintf("unable to blacklist token id %s: %s", tokenID, err.Error()))
+	}
+
+	// Add to cache only after token is blacklisted in database
+	ssas.TokenBlacklisted(ssas.Event{Op: "TokenBlacklist", TrackingID: tokenID, TokenID: tokenID})
+	b.tc.Set(tokenID, entryDate.Unix(), b.tcLifetime)
+
+	return nil
+}
+
+//	IsTokenBlacklisted tests whether this tokenID is in the blacklist cache.
+//	- Tokens should expire before blacklist entries, so a tokenID for a recently expired token may return "true."
+//	- This queries the cache only, so if a tokenID has been blacklisted on a different instance, it will return "false"
+//		until the cached blacklist is refreshed from the database.
+func (b *blacklist) IsTokenBlacklisted(tokenID string) bool {
+	b.RLock()
+	defer b.RUnlock()
+
+	bEvent := ssas.Event{Op: "TokenVerification", TrackingID: b.ID, TokenID: tokenID}
+	if _, found := b.tc.Get(tokenID); found {
+		ssas.BlacklistedTokenPresented(bEvent)
+		return true
+	}
+	return false
+}
+
+func (b *blacklist) BlacklistGroup(field GroupField, value string) error {
+	entryDate := time.Now()
+	expirationDate := entryDate.Add(b.gcLifetime)
+	entry, err := createGroupEntry(value, entryDate, expirationDate, field)
+	if err != nil {
+		return fmt.Errorf(fmt.Sprintf("unable to create group blacklist entry value %s: %s", value, err.Error()))
+	}
+
+	// Add to cache only after token is blacklisted in database
+	ssas.GroupBlacklisted(ssas.Event{Op: "GroupBlacklist", TrackingID: strconv.FormatUint(uint64(entry.ID), 10)})
+
+	b.gc.Set(uuid.New(), groupEntry{field, value}, b.gcLifetime)
+
+	return nil
+}
+
+func (b *blacklist) IsGroupBlacklisted(group ssas.Group) bool {
+	var cacheItems map[string]cache.Item
+
+	b.RLock()
+	cacheItems = b.gc.Items()
+	b.RUnlock()
+
+	var xDataObj map[string]interface{}
+	if err := json.Unmarshal([]byte(group.XData), &xDataObj); err != nil {
+		ssas.OperationFailed(ssas.Event{Op: "XDataUnmarshal", Help: err.Error()})
+	}
+
+	for _, item := range cacheItems {
+		entry, ok := item.Object.(groupEntry)
+		if !ok {
+			continue
+		}
+
+		switch entry.field {
+		case GroupFieldXData:
+			// Nothing to compare against
+			if xDataObj == nil {
+				continue
+			}
+			var expressionObj map[string]interface{}
+			if err := json.Unmarshal([]byte(entry.expression), &expressionObj); err != nil {
+				continue
+			}
+			if checkXData(xDataObj, expressionObj) {
+				ssas.BlacklistedGroupPresented(ssas.Event{Op: "GroupVerification", TrackingID: group.GroupID})
+				return true
+			}
+		}
+	}
+
+	return false
+}
+
+func (b *blacklist) startCacheRefreshTicker(refreshFreq time.Duration, refreshFunc func() error) (ticker *time.Ticker, done chan struct{}) {
+	event := ssas.Event{Op: "CacheRefreshTicker", TrackingID: b.ID}
+	ssas.ServiceStarted(event)
+
+	ticker = time.NewTicker(refreshFreq)
+	done = make(chan struct{})
+
+	go func() {
+		for {
+			select {
+			case <-done:
+				return
```

Comment:
To avoid leaking the go routine after we've stopped the ticker, we provide a signal channel that we'll use to tell the goroutine to exit. See [here](https://github.com/CMSgov/bcda-ssas-app/pull/56/files#diff-c641dc75064eafa5c9761e8a218e274f9984fd0b8e000ab3d5b8e473677e851aR264) for its usage.